File: expense-api/models/Category.php
<?php

class Category
{
    public static function findAll(): array
    {
        $db = Database::getConnection();
        $result = mysqli_query($db, 'SELECT id, nama_kategori, icon, created_at, updated_at FROM categories ORDER BY id ASC');
        return mysqli_fetch_all($result, MYSQLI_ASSOC);
    }

    public static function findById(int $id): ?array
    {
        $db = Database::getConnection();
        $stmt = mysqli_prepare($db, 'SELECT id, nama_kategori, icon, created_at, updated_at FROM categories WHERE id = ?');
        mysqli_stmt_bind_param($stmt, 'i', $id);
        mysqli_stmt_execute($stmt);
        $result = mysqli_stmt_get_result($stmt);
        $row = mysqli_fetch_assoc($result);
        mysqli_stmt_close($stmt);
        return $row ?: null;
    }

    public static function create(string $namaKategori, string $icon): array
    {
        $db = Database::getConnection();
        $stmt = mysqli_prepare($db, 'INSERT INTO categories (nama_kategori, icon) VALUES (?, ?)');
        mysqli_stmt_bind_param($stmt, 'ss', $namaKategori, $icon);
        mysqli_stmt_execute($stmt);
        $id = (int) mysqli_insert_id($db);
        mysqli_stmt_close($stmt);
        return self::findById($id);
    }

    public static function update(int $id, string $namaKategori, string $icon): ?array
    {
        $db = Database::getConnection();
        $stmt = mysqli_prepare($db, 'UPDATE categories SET nama_kategori = ?, icon = ? WHERE id = ?');
        mysqli_stmt_bind_param($stmt, 'ssi', $namaKategori, $icon, $id);
        mysqli_stmt_execute($stmt);
        mysqli_stmt_close($stmt);
        return self::findById($id);
    }

    public static function delete(int $id): bool
    {
        $db = Database::getConnection();
        $stmt = mysqli_prepare($db, 'DELETE FROM categories WHERE id = ?');
        mysqli_stmt_bind_param($stmt, 'i', $id);
        mysqli_stmt_execute($stmt);
        $affected = mysqli_affected_rows($db);
        mysqli_stmt_close($stmt);
        return $affected > 0;
    }
}

File: repository/category.go
package repository

import (
	"database/sql"
	"expense-api/config"
	"expense-api/models"
)

func GetAllCategories(userID int) ([]models.Category, error) {
	rows, err := config.DB.Query("SELECT id, nama_kategori, icon FROM categories WHERE user_id = ? ORDER BY id ASC", userID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var list []models.Category
	for rows.Next() {
		var c models.Category
		if err := rows.Scan(&c.ID, &c.NamaKategori, &c.Icon); err != nil {
			return nil, err
		}
		list = append(list, c)
	}
	return list, nil
}

func GetCategoryByID(id int) (*models.Category, error) {
	row := config.DB.QueryRow("SELECT id, nama_kategori, icon FROM categories WHERE id = ?", id)
	var c models.Category
	if err := row.Scan(&c.ID, &c.NamaKategori, &c.Icon); err != nil {
		return nil, err
	}
	return &c, nil
}

func CreateCategory(userID int, nama, icon string) (*models.Category, error) {
	result, err := config.DB.Exec("INSERT INTO categories (user_id, nama_kategori, icon) VALUES (?, ?, ?)", userID, nama, icon)
	if err != nil {
		return nil, err
	}
	id, _ := result.LastInsertId()
	return GetCategoryByID(int(id))
}

func UpdateCategory(id int, nama, icon string) (*models.Category, error) {
	_, err := config.DB.Exec("UPDATE categories SET nama_kategori = ?, icon = ? WHERE id = ?", nama, icon, id)
	if err != nil {
		return nil, err
	}
	return GetCategoryByID(id)
}

func DeleteCategory(id int) error {
	result, err := config.DB.Exec("DELETE FROM categories WHERE id = ?", id)
	if err != nil {
		return err
	}
	n, _ := result.RowsAffected()
	if n == 0 {
		return sql.ErrNoRows
	}
	return nil
}
